crates/matrix-sdk/src/authentication/service.rs

```diff
@@ -0,0 +1,566 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! High-level authentication APIs.
+#[cfg(feature = "experimental-oidc")]
+use std::collections::HashMap;
+use std::{fmt::Debug, path::PathBuf, sync::RwLock};
+
+#[cfg(feature = "experimental-oidc")]
+use mas_oidc_client::types::{
+    client_credentials::ClientCredentials,
+    errors::ClientErrorCode::AccessDenied,
+    registration::{ClientMetadata, VerifiedClientMetadata},
+    requests::Prompt,
+};
+#[cfg(feature = "experimental-oidc")]
+use ruma::api::client::discovery::discover_homeserver::AuthenticationServerInfo;
+use ruma::api::client::session::get_login_types;
+use url::Url;
+
+#[cfg(feature = "experimental-oidc")]
+use crate::oidc::{
+    registrations::{ClientId, OidcRegistrations, OidcRegistrationsError},
+    AuthorizationResponse, Oidc, OidcAuthorizationData, OidcError,
+};
+use crate::{sanitize_server_name, Client, ClientBuilder, IdParseError};
+
+/// A high-level service for authenticating a user with a homeserver.
+/// TODO: Add an example code snippet.
+pub struct AuthenticationService {
+    /// The base path to use for storing data. This could be the same path used
+    /// for the client, or the client might be placed within a subdirectory of
+    /// the base path.
+    pub base_path: PathBuf,
+    /// The user agent sent when making requests to the homeserver.
+    pub user_agent: Option<String>,
+    /// The client used to make requests to the homeserver.
+    client: RwLock<Option<Client>>,
+    /// Details about the currently configured homeserver.
+    homeserver_details: RwLock<Option<HomeserverLoginDetails>>,
+    /// Metadata about the current client that will be used for dynamic OIDC
+    /// client registration and likely shown to the user during login.
+    #[cfg(feature = "experimental-oidc")]
+    oidc_client_metadata: Option<ClientMetadata>,
+    /// Any static client registrations that have been made for this client
+    /// against particular authentication issuers.
+    #[cfg(feature = "experimental-oidc")]
+    oidc_static_registrations: HashMap<Url, ClientId>,
+    /// A sliding sync proxy URL that will override any proxy discovered from
+    /// the homeserver. Setting this value allows you to use a proxy against a
+    /// homeserver that hasn't yet been configured with one.
+    #[cfg(feature = "experimental-sliding-sync")]
+    pub custom_sliding_sync_proxy: RwLock<Option<String>>,
+}
+
+impl Debug for AuthenticationService {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        f.debug_struct("AuthenticationService")
+            .field("homeserver_details", &self.homeserver_details)
+            .finish()
+    }
+}
+
+/// Errors related to authentication through the `AuthenticationService`.
+#[derive(Debug, thiserror::Error)]
+#[cfg_attr(feature = "uniffi", derive(uniffi::Error), uniffi(flat_error))]
```

Comment:
I have a feeling the right option here would (sadly) be to duplicate this enum in the FFI, and box the underlying messages instead of using `to_string()` on them. Didn't want to do that until this had been reviewed, maybe its ok, I'm not sure?